### ISSUE TITLE
A better literate documentation mechanism

### ISSUE THREAD

**Issue by rix0rrr** [aws/jsii#391]
The literate documentation system we have today (where we include a piece of a file in a subdirectory) is not good enough. It requires a lot of overhead, making the additional file for every example you want to include and then referencing it. The overhead makes people skip it for small one-line examples, which makes it miss the point (because the one-line examples inevitably go out of date). This issue describes mechanism which will both lower the bar to entry, and make it impossible to write uncompiled code samples.

I will describe it using an example, which I think will be more clear than a lot of prose:

# Current mechanism

* `test/integ.some-example.lit.ts`

```ts
import mylib = require('../lib');

class SomeStack extends cdk.Stack {
  constructor(...) {
     super(this, id);
    
     ///! show
     const myconstruct = new mylib.Construct(this, 'hello');
     // relevant example goes here
    /// !hide
  }
}

const app = new cdk.App();
new SomeStack(app, 'mystack');
app.run();
```

* `README.md`

```
Welcome to the readme for my library!

Here is how you use it:

[example of using my library](test/integ.some-example.lit.ts)

As you can see above, ...
```

# Proposed mechanism

* `test/docfixture.default.tpl`

```ts
import mylib = require('../lib');

class SomeStack extends cdk.Stack {
  constructor(...) {
     super(this, id);
    
     /* HERE */
  }
}

const app = new cdk.App();
new SomeStack(app, 'mystack');
app.run();
```
* `README.md`

```
Welcome to the readme for my library!

Here is how you use it:

` ` `ts
const myconstruct = new mylib.Construct(this, 'hello');
// relevant example goes here
` ` `

If you want to integrate with another resource:

` ` `ts
/// <prereq>
// This won't be displayed in the README but it will be compiled.
const other = new otherlib.Resource();
/// </prereq>

const myconstruct = new mylib.Construct(this, 'hello');
myconstruct.connectTo(other);
` ` `

```

# Bottom line

* We will insert all MarkDown blocks fenced by `` ```ts `` into a template and the resulting code must compile.
* Additional instructions (using a different template than the default, extra lines of code just to make this example work) can be given via special comment syntax, which also won't end up in the final code. Proposal: `///`.
* It is not allowed to have code blocks in a README and NOT have a fixture file to compile them against. That is: all code must be compile-checked.
* Implemented by jsii.

**Comment by rix0rrr** [comment]
Extensions:

We should probably extend the same courtesies to code blocks encountered in doc comments.

**Comment by eladb** [comment]
I am wondering if it would be nicer to simple compile the readme.md file during build (and check in the result), so we can have the readme completely self contained (with fixtures and all).

**Comment by Doug-AWS** [comment]
Will we add the compile step to checkin? I.e., one cannot check in a file with a code example that does not compile?

**Comment by rix0rrr** [comment]
> I am wondering if it would be nicer to simple compile the readme.md file during build (and check in the result), so we can have the readme completely self contained (with fixtures and all).

I am not sure what you mean by "compiling the readme", or what checking in the result means, but I did plan on applying the mechanism describe above during the jsii build.

**Comment by eladb** [comment]
What I meant is that maybe there's a `README.md.src` that includes both markdown and also code blocks with directives (`//!`) to the renderer:

```md
Let's define our first CDK stack.

'''sh
//!exec
//!hide
npm i @aws-cdk/cdk, @aws-cdk/aws-sqs
'''

Create a file `stack.ts`:

'''ts
//!new-file:stack.ts
import { Stack } from '@aws-cdk/cdk';

export class MyStack extends Stack { }
'''

Now, create a file `app.ts`:

'''ts
//!new-file:app.ts
import { App } from '@aws-cdk/cdk';
import { MyStack } from './stack';

const app = new App();
new MyStack(app, 'MyFirstStack');
'''

'''sh
//!exec
//!hide
tsc
'''

Now, let's add a resource to our stack:

'''ts
//!edit-file:stack.ts
//!delete-lines:4-4
export class MyStack extends Stack {
  constructor(..) {
    super();
    new sqs.Queue(this, 'MyQueue');
  }
}
'''
```

YMMV of course... Just made up those directives...

Then, we have a little renderer that reads this `README.md.src`, executes all the directives and produces a `README.md` file as a result without directives or hidden stuff.

The nice thing here is that we can do pretty crazy things and generate portions of the readme, markdown, common headers and footers, etc.

We can decide if we check in the READM.md file. Not required. Eventually the important thing is that the README in npmjs, pypi and our docs website is the rendered version.

**Comment by rix0rrr** [comment]
At first glance, I hate this (all the annotation work!) and I shudder to think about the security implications of letting people run arbitrary shell commands on our boxes just by submitting a PR :).

But those are details, and it occurs to me that you're trying to solve a slightly different problem than I am.

I'm interested in having code snippets appear in the docs syntax-checked. The snippets just appear by themselves in either the README or the doc blocks of methods and classes. That really doesn't need much more than a project template and some holes. It doesn't need sequencing, it doesn't need file editing, it doesn't need arbitrary package installs.

You seem to be solving the problem of a walkthrough (a la the getting started guide, or the workshop site), where we're taking a user by the hand, telling them "do this now, do that now, run this command, modify this file you created previously to add these lines, etc.", and having that validated to make sure it's still correct.

That problem is more generic, and will need more flexibility. We can probably combine the two in some way.

**Comment by rix0rrr** [comment]
I don't `README.md.src` is necessary.

If we did that, we wouldn't have a README visible on GitHub, which is not great, or we would have to commit the generated README.md (also not great).

That decisions doesn't need to impact what gets published to NPM or PyPI, those markdowns could be post-processed, loaded from the JSII Assembly.

**Comment by RomainMuller** [comment]
While @eladb's solutions strikes me as more flexible, it also feels a little over-complicated to me. I also worry about it being kind of brittle (edit commands for example, are quite brittle).
